content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md

```diff
@@ -0,0 +1,308 @@
+
+#### Closing Threads
+In closing threads, the following functions are used.
+
+```C++
+pthread_exit (status) 
+```
+
+`Pthread_exit()` terminates a thread. In most cases, a thread's duty is completed using `pthread_exit().` This method allows threads to continue running after the `main()` function. If not, they will be stopped after the main() method.
+
+Here is a C++ application to illustrate the two actions.
+
+```C++
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+void * WelcomeMessage(void * info)
+{
+   int information = (int)info;  
+
+   printf("\n Hi and welcome to Section Engineering education - got %d !\n", information);
+   pthread_exit(NULL);
+}
+
+int main()
+{
+   int xy;
+   pthread_t thread_id;
+
+   int b = 6;
+
+   xy = pthread_create(&thread_id, NULL, WelcomeMessage, (void*)b);
+   if(xy)
+   {
+    printf("\n ERROR: An error message from pthread_create is %d \n", xy);
+    exit(1);
+   }
+   printf("\n A new thread was developed (%u)... \n", thread_id);
+
+   pthread_exit(NULL);
+}
+```
+
+The output of the code will be:
+
+```bash
+A new thread was developed (3179116288)... 
+
+Hi and welcome to Section Engineering education - got 11 !
+```
+
+To run the above code in the UNIX system, run the below command:
+
+```bash
+gcc first.c -lpthread -o output1
+```  
+
+GCC is the software used to run the C++ code in Linux systems.
+The `-lpthred` command tells the command that the compiled file is a pthread program. 
+The `-o' specifies the output of the compiled program.
+
+After the code is compiled, an output file is created and saved in the terminal's same directory. For our case, the output file will be saved as `output1` as we specified in the command. To view the output, run the below code:
+
+```bash 
+./output1
+```
+
+The below screenshot shows all the processes involved.
+
+![Process illustration](/engineering-education/understanding-shared-memory-programming-with-pthreads-and-openmp/process.png)
+
+#### Cancelling a thread
+Assuming everything goes well, a thread will end by executing `pthread_exit` or returning from its thread function. A thread may, however, request that another thread be terminated. This is referred to as canceling a thread. Use `pthread_cancel` with the thread ID to cancel a thread. Unless a thread is detached, a canceled thread may be rejoined to free up resources. A thread may be generated as a joinable (default) or detached (optional). A joinable thread is not automatically cleaned away by GNU/Linux as a process. Instead, the thread's exit state is stored until another thread performs pthread join to retrieve it. Then its resources are freed. A detached thread, on the other hand, cleans up automatically. Because a detached thread is quickly cleaned up, another thread cannot synchronize or acquire its return value via pthread join.

```

Comment:
This is a very long sentence. Consider breaking it into smaller sentences with independent meanings.